pageserver/src/basebackup.rs

```diff
@@ -252,6 +255,32 @@ where
     async fn send_tarball(mut self) -> Result<(), BasebackupError> {
         // TODO include checksum
 
+        // On normal Postgres shutdown, shutdown checkpoint should be performed and correspondent
+        // WAL record should be the last record in the WAL.
+        let normal_shutdown = if let Ok(checkpoint_bytes) =
+            self.timeline.get_checkpoint(self.lsn, self.ctx).await
+        {
+            let checkpoint =
+                CheckPoint::decode(&checkpoint_bytes).context("deserialize checkpoint")?;
+            let checkpoint_end =
+                checkpoint.redo + ((XLOG_SIZE_OF_XLOG_RECORD + 8 + SIZEOF_CHECKPOINT) as u64);
+            let checkpoint_end_lsn = Lsn(checkpoint_end
+                + (if (XLOG_BLCKSZ as u64) - checkpoint.redo % (XLOG_BLCKSZ as u64)
+                    < (SIZEOF_CHECKPOINT as u64)
+                {
+                    if (checkpoint_end & ((WAL_SEGMENT_SIZE - 1) as u64)) < (XLOG_BLCKSZ as u64) {
+                        XLOG_SIZE_OF_XLOG_LONG_PHD
+                    } else {
+                        XLOG_SIZE_OF_XLOG_SHORT_PHD
+                    }
+                } else {
+                    0
+                }) as u64);
```

Comment:
Let's create a helper function for this. Something like:

```
/// Given the start LSN of a WAL record and its length, calculate its end LSN.
///
/// In the simple cases, the end LSN is just `start_lsn + len`, but it's more complicated if 
/// the record crosses a page boundary.
///
/// `len` is the length not including any WAL page headers. It should match the `xl_tot_len`
/// field on the WAL record.
fn calculate_walrecord_end_lsn(start_lsn: Lsn, len: u32) -> Lsn

Reply:
Done